### ISSUE TITLE
suggest versioned releases of the FMS package...

### ISSUE THREAD

**Issue by edhartnett** [NOAA-GFDL/FMS#102]
@underwoo , @bensonr, @Zhi-Liang and other FMS code leaders, I suggest that you do a versioned release of this package immediately, and then do them periodically thereafter.

I understand that you also have a larger versioning scheme, which applies to your entire model stack, in which you tag every repo involved to mark the exact software used in a particular model build. This meets the versioning needs of the users of that model, but not of other FMS users (like me).

What I am suggesting is that, _in addition to that_, you release versions of FMS, so that external users can just install a version of FMS and use it until its next release.

If you do a versioned release, the current master and all branches are all still available for anyone using them in their workflow. The ability to create an over-all tag for all software is still there, and all your existing workflows still work.

But new workflows, in accordance with general practice, can be followed. Users and researchers will have a canonical version number for the FMS package. Different FMS users will be able to understand what version of the software they are using, and be able to manage upgrades properly.

For example, when I add PIO capability to FMS, I will be testing it with various versions of netcdf-c, netcdf-fortran, parallel-netcdf, and HDF5. This is so I can say: "version 2.0 of FMS has been tested with netcdf-c 4.6.2, 4.6.3, etc." Without this information, how do sysadmins know what packages they need to upgrade to support your software?

Providing a well-understood versioned release is part of having a library that others are relying upon. It's necessary so we all can manage our dependencies.

To do a versioned release of FMS, there are a few quick steps that must be followed. It's about a 5-minute process to generate a versioned, tested, portable, and stand-alone tarball for the FMS package. 

Please let me know your thinking, and I can guide you through the process the first time. In particular, there are some shared library version-info numbers that must be updated properly, so that shared libraries will work with this code correctly.

**Comment by edwardhartnett** [comment]
Guys, can we do a versioned release of FMS? I can show you how in a few minutes.

Doing versioned releases will not interfere with your home-rolled build and deployment system (i.e. your "workflow"). But it will help external users like myself, and will begin to support the idea of GFDL eventually moving to a more industry-normal way of doing release and deployment.

We could do this today - it takes about 5 minutes, for whomever is the admin of this repo.

**Comment by wrongkindofdoctor** [comment]
We will be implementing versioned releases of FMS and FMScoupler after we have finished testing the latest round of FMS commits with the supported model configurations.

**Comment by edwardhartnett** [comment]
OK, awesome. Tell me what version number or string you want to use, and I will put up a PR with the changes to configure to support that.

Also. each time you do a versioned release, you need to update the version info in the shared library build. There are web instructions for how to update this - it is not something you should update without understanding, but it must be updated each release if shared libraries are to work properly.

**Comment by edwardhartnett** [comment]
Guys, are we getting any closer to this day?

You are awaiting perfection, and that is a mistake. Release a version immediately, and the next version will be more perfect. Waiting for perfection is just the same as procrastinating and putting off the process changes you need to make.

**Comment by underwoo** [comment]
We are not waiting for perfection, but for a time we feel it is time to do a release.  We are still working through bugs and tests for the current code before creating a new release.

**Comment by edwardhartnett** [comment]
This is already operational code, right?

So you have done a release. The release is the operational code, with whatever way that you have tagged it.

What I am requesting is that you label that release with a version number, and generate and distribute a release tarball. (This can be in addition to your proprietary tagging system that you use for internal purposes.)

OK, I see I am pushing perhaps too hard, too fast. Let us see if we can get some testing working, and then move on to the issue of a release later.

**Comment by underwoo** [comment]
We have released [2019.01](../tree/2019.01) that has an updated `configure.ac` with the release version number.  The master branch has also been updated with the next version expected for release.  We'll work to keep the version number in `configure.ac` up to date with development and releases.

**Comment by edwardhartnett** [comment]
Great! What is done on the netCDF project is to now change the version to 2019.01-development (that is, add "-development" to the configure.ac version number on master).

Then, anyone building from master will get a different version than someone building from releases.

Before the next release, you change the version number again to the new release, and then, once that's done, add "-development" again.

Also, for future FMS releases, you need to change the version-info setting for libtool. This is in libFMS/Makefile.am:

```
# These linker flags specify libtool version info.
# See http://www.gnu.org/software/libtool/manual/libtool.html#Libtool-versioning
# for information regarding incrementing `-version-info`.
libFMS_la_LDFLAGS = -version-info 2:0:0
```

This is not a version number that you set, you must increment the three parts of version-info in accordance with the rules described in the manual (see the link above). What do you depends on whether you have added, removed, or changed functions since the last release.
